src/rust/index.rs

```diff
@@ -0,0 +1,413 @@
+    protein_info: HashMap<String, KmerSignature>,
+    moltype: String,
+    ksize: u32,
+    scaled: u32,
+}
+
+pub struct ProteomeIndex {
+    // RocksDB instance for persistent storage
+    db: DB,
+
+    // Collection of all signatures for searching
+    collection: Arc<Mutex<Collection>>,
+
+    // Combined minhash of all proteins for statistics
+    combined_minhash: Arc<Mutex<KmerMinHash>>,
+
+    // Map of signature md5 -> protein k-mer information
+    protein_info: Arc<Mutex<HashMap<String, Protein>>>,
+
+    // Amino acid ambiguity handler
+    aa_ambiguity: Arc<AminoAcidAmbiguity>,
+
+    // Protein encoding function
+    encoding_fn: fn(u8) -> u8,
+
+    // Add moltype field for serialization
+    moltype: String,
+
+    // Add ksize field for serialization
+    ksize: u32,
+
+    // Add scaled field for serialization
+    scaled: u32,
+
+    // Statistics for k-mer frequencies and IDF
+    stats: ProteomeIndexKmerStats,
+}
+
+impl ProteomeIndex {
+    pub fn new<P: AsRef<Path>>(path: P, ksize: u32, scaled: u32, moltype: &str) -> Result<Self> {
+        // Create RocksDB options
+        let mut opts = Options::default();
+        opts.create_if_missing(true);
+
+        // Open the database
+        let db = DB::open(&opts, path)?;
+
+        let hash_function = match get_hash_function_from_moltype(moltype) {
+            Ok(value) => value,
+            Err(value) => return value,
+        };
+
+        let encoding_fn = match get_encoding_fn_from_moltype(moltype) {
+            Ok(value) => value,
+            Err(value) => return value,
+        };
+
+        // Create the minhash sketch
+        let minhash = KmerMinHash::new(
+            scaled,
+            ksize,
+            hash_function,
+            42,   // seed
+            true, // track_abundance
+            0,    // num (use scaled instead)
+        );
+
+        // Create an empty collection with storage
+        let manifest = Manifest::default();
+        let storage =
+            InnerStorage::new(FSStorage::builder().fullpath("".into()).subdir("".into()).build());
+        let collection = Collection::new(manifest, storage);
+
+        Ok(Self {
+            db,
+            collection: Arc::new(Mutex::new(collection)),
+            combined_minhash: Arc::new(Mutex::new(minhash)),
+            protein_info: Arc::new(Mutex::new(HashMap::new())),
+            aa_ambiguity: Arc::new(AminoAcidAmbiguity::new()),
+            encoding_fn,
+            moltype: moltype.to_string(),
+            ksize,
+            scaled,
+            stats: ProteomeIndexKmerStats { idf: HashMap::new(), frequency: HashMap::new() },
+        })
+    }
+
+    pub fn add_uniprot_entry(&mut self, uniprot_entry: UniProtEntry) -> Result<()> {
+        // Get parameters from combined_minhash
+        let mut minhash_guard = self.combined_minhash.lock().unwrap();
+        let ksize = minhash_guard.ksize() as u32;
+        let scaled = minhash_guard.scaled();
```

Comment:
```suggestion
        let (ksize, scaled) = {
            let guard = self.combined_minhash.lock().unwrap();
            (guard.ksize() as u32, guard.scaled())
        };
```

This would take profiling -- the suggestion (in combination with the write, below, to update the min_hash) is to acquire the lock, read `ksize` and `scaled`, and then drop it. Do the processing, and then acquire the lock again to do the writing, rather then keeping the lock while tasks that don't require it occur. It also increases the number of times the lock needs to be acquired -- might not be ideal.

I find this easier to read, though -- easier to see when the lock is actually necessary

Reply:
And, indeed, it's quite a shame to trigger a lock simply to read some parameters of the minhash. I'd recommend thinking about storing these in `ProteomeIndex` to avoid the lock at all.